subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java

```diff
@@ -18,42 +18,43 @@
 
 import com.google.common.primitives.Ints;
 
-import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
 
 /**
  * An immutable hash code. Must be 4-255 bytes long.
  * Inspired by the Google Guava project – https://github.com/google/guava.
  */
-public class HashCode implements Serializable, Comparable<HashCode> {
+public abstract class HashCode implements Serializable, Comparable<HashCode> {
```

Comment:
Creating subclasses is likely to degrade performance, as the JIT can no longer inline if multiple implementations are used. If our hashcodes are always 128bit, we should change the implementation and not allow any other size.

Reply:
Actually the JIT can detect whether the others are unused and eliminate the rest and treat this as the only implementation, it won't become bi- or megamorphic.
This is confirmed by the attached JMH test, the speed of having a single impl or multiple, where only one is used is the same.

Reply:
Yes, that's why I said we should make it the only implementation if we don't need anything else. Otherwise it's too easy to call the other impl once and ruin the optimization.

Reply:
It may be possible to use SHA1 also instead of MD5 currently, but I'm not aware of it being used.
I don't mind either way.